### PR TITLE
Alex lets terminate

### PR DESCRIPTION
**Pull request by alek-thunder**
Finally done. After heavy refactoring the pipelines and all concurrent steps terminate now properly.

### CODE REVIEW

File: bitflow/io/sinksteps.py [hunk 1]
```diff
@@ -130,17 +111,23 @@ def read(self, packet_size):
...
         super().__init__()
         self.__name__ = "ListenSink"
-        self.marshaller = get_marshaller_by_data_format(data_format)
+        self.threaded_step = _ListenSink(self.sample_queue_in, self.sample_queue_out, self.input_counter,
+                                         host, port, data_format, sample_buffer_size, max_receivers, retry_on_close)
+
+
+class _ListenSink(_AsyncProcessingStep):
 
+    def __init__(self, sample_queue_in, sample_queue_out, input_counter, host: str = "0.0.0.0", port: int = 5010,
```
Comment (mwallschlaeger) [inline bitflow/io/sinksteps.py]: defaults already set in outer class. remove or define GLOBAL

File: bitflow/processingstep.py [hunk 2]
```diff
@@ -146,70 +153,182 @@ def initialize_step(name, script_args):
...
 
-class ProcessingStep:
+class ProcessingStep(metaclass=helper.CtrlMethodDecorator):
     """ Abstract ProcessingStep Class"""
-    subclasses = []
     __description__ = "No description written yet."
 
     def __init__(self):
+        self.__name__ = "ProcessingStep"
+        self.lock = multiprocessing.Lock()
```
Comment (mwallschlaeger) [inline bitflow/processingstep.py]: this does not inherit from multiprocessing, why is this necessary here? Sinksteps defining their own threading.lock

File: bitflow/io/sources.py [hunk 3]
```diff
@@ -24,13 +33,52 @@ def read_header(marshaller, s, buffer_size=2048):
...
+        self.running.value = 0
+
+
+class _Source(multiprocessing.Process, metaclass=helper.CtrlMethodDecorator):
 
-    def __init__(self, queue, marshaller, running):
+    def __init__(self, pipeline, marshaller, running):
+        self.pipeline = pipeline
+        self.pipeline_sample_queue_in = pipeline.sample_queue_in
+        self.pipeline_sample_queue_out = pipeline.sample_queue_out
```
Comment (mwallschlaeger) [inline bitflow/io/sources.py]: Is this necessary? Could pipeline just provide a int value with out_samples? Or what else is this queue used for, than just count the number of out samples?

File: bitflow/pipeline.py [hunk 4]
```diff
@@ -1,164 +1,149 @@
...
+        super().__init__()
         self.__name__ = "Pipeline"
-        self.queue = None
+        self.lock = multiprocessing.Lock()
+        self.input_counter = multiprocessing.Value('i', 0)
         self.maxsize = maxsize
-        self.queue = self.create_queue(multiprocessing_input)
-        self.subqueues = {}
+        self.sample_queue_in = self.create_queue(multiprocessing_input)
+        self.sample_queue_out = self.create_queue(multiprocessing_input)
```
Comment (mwallschlaeger) [inline bitflow/pipeline.py]: Why is this pl needed?

File: bitflow/io/sources.py (hunk 3 above)
Comment (alek-thunder) [reply bitflow/io/sources.py]: I get your point but in the current setup it is required because the destination pipeline of a source does not "know" that its predecessor is a "source". It will read from queue_in and write to queue_out. Without reading the queue_out at source, it will overfill and block. Will also add an join() in on_close to correctly close the queue.

I plan to rework this in the next PR to make it more consistent with other parallel things like batch and fork.

File: bitflow/pipeline.py (hunk 4 above)
Comment (alek-thunder) [reply bitflow/pipeline.py]: Pipelines can be within steps (like forks and batches) the queue_in is used to get samples into the parallel running pipe. The queue out is used to return results to the outer pipe for further processing. Its like:

Outer Pipe process ------------------------------------------------->
                       queue_in ----->|                        |<----queue_out
Inner pipe process                  --------------------
